Factor out start-a-libp2p-daemon!

AlexKnauth opened the pull request:
- `start-libp2p-daemon!` for "a" daemon
 - `stop-libp2p-daemon!` for "a" daemon
 - `start-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`
 - `stop-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`

See also https://github.com/vyzo/gerbil-libp2p/issues/20

AlexKnauth committed:
Factor out start-a-libp2p-daemon!

to avoid current-libp2p-daemon problems

bennewhall reviewed (commented):
If backwards compatibility is not an issue, I am personally in favor of removing current-libp2p-daemon entirely. I personally have not found much use for it in my own code since start-libp2p-daemon! returns a daemon. As far as the code goes LGTM

fare commented on libp2p.ss:
```diff
@@ -14,7 +14,9 @@
   start-libp2p-daemon!
+  start-a-libp2p-daemon!
   stop-libp2p-daemon!
+  stop-the-libp2p-daemon!
```
Why is it start-a- and stop-the- ?

AlexKnauth replied:
If backwards compatibility is a concern,
Current behavior has:
 - `start-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`
 - `stop-libp2p-daemon!` for "a" daemon, where `current-libp2p-daemon` doesn't matter

A better design would have:
 - `start-libp2p-daemon!` for "a" daemon
 - `stop-libp2p-daemon!` for "a" daemon
 - `start-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`
 - `stop-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`

The closest thing I could think of to that while maintaining backwards compatibility has:
 - `start-a-libp2p-daemon!` for "a daemon"
 - `stop-libp2p-daemon!` for "a" daemon as dictated by existing code
 - `start-libp2p-daemon!` for "the" daemon as dictated by existing code
 - `stop-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`

If backwards compatibility is not a concern, I could make it more consistent with only "the" and no "a", but I won't do that unless @vyzo says that's what he wants.

AlexKnauth replied:
Okay, I've changed the convention to make it more consistent with only "the" and no "a", and also added some documentation.

fare commented on libp2p/daemon.ss:
```diff
@@ -14,6 +14,8 @@
   (make-parameter #f))
 
+;; Starts a new libp2p-daemon only if there is no existing current-libp2p-daemon,
+;; and if so sets the current-libp2p-daemon to the new one.
```
Maybe that should be renamed `ensure-libp2p-daemon` ?
Or, if we want backward-compatibility, a `force:` or `new:` boolean flag?

fare commented on libp2p/daemon.ss:
```diff
@@ -45,5 +58,9 @@
 
+(def (stop-the-libp2p-daemon!)
+  (stop-libp2p-daemon!)
+  (current-libp2p-daemon #f))
```
Shouldn't `stop-libp2p-daemon!` always call `(current-libp2p-daemon #f)` ? And then this function becomes unnecessary?

fare reviewed (commented):
I don't like your proposed API changes as is. Or at least, you'll have to justify them.

vyzo commented:
Honestly, i am fine either way -- i'll let @fare drive this.

AlexKnauth commented:
Okay, I think I'll rename them to fit this convention:
 - `start-libp2p-daemon!` for "a" daemon
 - `stop-libp2p-daemon!` for "a" daemon
 - `start-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`
 - `stop-the-libp2p-daemon!` for "the" daemon in `current-libp2p-daemon`

AlexKnauth committed:
Rename start-libp2p-daemon -> start-the-libp2p-daemon!

convention: names with "the" libp2p-daemon convention for functions that deal with current-libp2p-daemon

AlexKnauth committed:
Document new libp2p-daemon functions
